orchagent/muxorch.cpp

```diff
@@ -0,0 +1,991 @@
+
+/* Global variables */
+extern Directory<Orch*> gDirectory;
+extern CrmOrch *gCrmOrch;
+extern NeighOrch *gNeighOrch;
+extern AclOrch *gAclOrch;
+extern PortsOrch *gPortsOrch;
+
+extern sai_object_id_t gVirtualRouterId;
+extern sai_object_id_t  gUnderlayIfId;
+extern sai_object_id_t gSwitchId;
+extern sai_route_api_t* sai_route_api;
+extern sai_tunnel_api_t* sai_tunnel_api;
+extern sai_next_hop_api_t* sai_next_hop_api;
+extern sai_router_interface_api_t* sai_router_intfs_api;
+
+/* Constants */
+#define MUX_TUNNEL "MUX_TUNNEL0"
+#define MUX_ACL_TABLE_NAME "mux_acl_table";
+#define MUX_ACL_RULE_NAME "mux_acl_rule";
+#define MUX_HW_STATE_UNKNOWN "unknown"
+#define MUX_HW_STATE_PENDING "pending"
+
+const map<std::pair<MuxState, MuxState>, MuxStateChange> muxStateTransition =
+{
+    { { MuxState::MUX_STATE_INIT, MuxState::MUX_STATE_ACTIVE}, MuxStateChange::MUX_STATE_INIT_ACTIVE
+    },
+
+    { { MuxState::MUX_STATE_INIT, MuxState::MUX_STATE_STANDBY}, MuxStateChange::MUX_STATE_INIT_STANDBY
+    },
+
+    { { MuxState::MUX_STATE_ACTIVE, MuxState::MUX_STATE_STANDBY}, MuxStateChange::MUX_STATE_ACTIVE_STANDBY
+    },
+
+    { { MuxState::MUX_STATE_STANDBY, MuxState::MUX_STATE_ACTIVE}, MuxStateChange::MUX_STATE_STANDBY_ACTIVE
+    },
+};
+
+const map <MuxState, string> muxStateValToString =
+{
+    { MuxState::MUX_STATE_ACTIVE, "active" },
+    { MuxState::MUX_STATE_STANDBY, "standby" },
+    { MuxState::MUX_STATE_INIT, "init" },
+    { MuxState::MUX_STATE_FAILED, "failed" },
+    { MuxState::MUX_STATE_PENDING, "pending" },
+};
+
+const map <string, MuxState> muxStateStringToVal =
+{
+    { "active", MuxState::MUX_STATE_ACTIVE },
+    { "standby", MuxState::MUX_STATE_STANDBY },
+    { "init", MuxState::MUX_STATE_INIT },
+    { "failed", MuxState::MUX_STATE_FAILED },
+    { "pending", MuxState::MUX_STATE_PENDING },
+};
+
+static inline MuxStateChange mux_state_change (MuxState prev, MuxState curr)
+{
+    auto key = std::make_pair(prev, curr);
+    if (muxStateTransition.find(key) != muxStateTransition.end())
+    {
+        return muxStateTransition.at(key);
+    }
+
+    return MuxStateChange::MUX_STATE_UNKNOWN_STATE;
+}
+
+static sai_status_t create_route(IpPrefix &pfx, sai_object_id_t nh)
+{
+    sai_route_entry_t route_entry;
+    route_entry.switch_id = gSwitchId;
+    route_entry.vr_id = gVirtualRouterId;
+    copy(route_entry.destination, pfx);
+    subnet(route_entry.destination, route_entry.destination);
+
+    sai_attribute_t attr;
+    vector<sai_attribute_t> attrs;
+
+    attr.id = SAI_ROUTE_ENTRY_ATTR_PACKET_ACTION;
+    attr.value.s32 = SAI_PACKET_ACTION_FORWARD;
+    attrs.push_back(attr);
+
+    attr.id = SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID;
+    attr.value.oid = nh;
+    attrs.push_back(attr);
+
+    sai_status_t status = sai_route_api->create_route_entry(&route_entry, (uint32_t)attrs.size(), attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to create tunnel route %s, rv:%d",
```

Comment:
To ease debug you may want to consider also capture nh oid in the syslog when create route failed.

Reply:
Addressed